### PR TITLE
Add advanced fuzzer

### PR DESCRIPTION
**Pull request by maxammann**
Adds a fuzzer which:
* Can call contracts
* Uses complete scripts with fuzzed script data
* Can use LibAFL

## Manual

```
cargo install cargo-fuzz
apt install clang pkg-config libssl-dev # for LibAFL
rustup component add llvm-tools-preview --toolchain nightly
```

### Generate Seeds

It is necessary to first convert Sway programs into a suitable format for use as seed input to the fuzzer. This can be done with the following command:
```
cargo run --bin seed <input dir> <output dir>
```

### Running the Fuzzer
The Rust nightly version is required for executing cargo-fuzz. We also disable AddressSanitizer for a significant speed improvement, as we do not expect memory issues in a Rust program that does not use a significant amount of unsafe code, which our [cargo-geiger](https://github.com/rust-secure-code/cargo-geiger) analysis showed. It makes sense to leave AddressSanitizer turned on if the Fuel project uses more unsafe Rust in the future (either directly or through dependencies). The remaining flags are either required for LibAFL or are useful to make it use seven cores.
```
cargo +nightly fuzz run --sanitizer none grammar_aware -- \
	-ignore_crashes=1 -ignore_timeouts=1 -ignore_ooms=1 -fork=7
```

If you use libfuzzer (default) then the following command is enough:

```
cargo fuzz run --sanitizer none grammar_aware
```

### Execute a Test Case
Test cases can be executed using the following command. This is useful for triaging issues.
```
cd fuzz/
cargo run --bin execute <file/dir>
```

### Collect Statistics
We created a tool that writes gas statistics to a file called gas_statistics.csv. This can be used to analyze the execution time versus gas usage on a test corpus.
```
cargo run --bin collect <file/dir>
```

### Generate Coverage
Regardless of how inputs are generated, it is important to measure a fuzzing campaign’s coverage after its run. To perform this measure, we used the support provided by cargo-fuzz and [rustc](https://doc.rust-lang.org/stable/rustc/instrument-coverage.html). First, install [cargo-binutils](https://github.com/rust-embedded/cargo-binutils#installation). After that, execute the following command:
```
cargo +nightly fuzz coverage grammar_aware corpus/grammar_aware
```
Finally, generate an HTML file using LLVM:

```
cargo cov -- show
target/x86_64-unknown-linux-gnu/coverage/x86_64-unknown-linux-gnu/release/grammar_aware --format=html -instr-profile=coverage/grammar_aware/coverage.profdata /root/audit/fuel-vm > index.html
```

The last argument to `cargo cov` defines the filter based on the host file paths.

## Checklist
- [x] Breaking changes are clearly marked as such in the PR description and changelog
- [x] New behavior is reflected in tests
- [x] If performance characteristic of an instruction change, update gas costs as well or make a follow-up PR for that
- [x] [The specification](https://github.com/FuelLabs/fuel-specs/) matches the implemented behavior (link update PR if changes are needed)

### Before requesting review
- [x] I have reviewed the code myself
- [x] I have created follow-up issues caused by this PR and linked them here

### After merging, notify other teams

TODO

- [ ] [Rust SDK](https://github.com/FuelLabs/fuels-rs/)
- [ ] [Sway compiler](https://github.com/FuelLabs/sway/)
- [ ] [Platform documentation](https://github.com/FuelLabs/devrel-requests/issues/new?assignees=&labels=new+request&projects=&template=NEW-REQUEST.yml&title=%5BRequest%5D%3A+) (for out-of-organization contributors, the person merging the PR will do this)
- [ ] Someone else?

### CODE REVIEW

File: fuel-asm/src/macros.rs [hunk 1]
```diff
@@ -1007,6 +1007,7 @@ macro_rules! impl_instructions {
         /// Solely the opcode portion of an instruction represented as a single byte.
         #[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
         #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+        #[cfg_attr(feature = "arbitrary", derive(arbitrary::Arbitrary))]
```
Comment (maxammann) [inline fuel-asm/src/macros.rs:1010]: This is required for the previous fuzzing harness, which I fixed.

File: fuel-crypto/src/secp256/signature_format.rs [hunk 2]
```diff
@@ -55,6 +55,8 @@ impl TryFrom<secp256k1::ecdsa::RecoveryId> for RecoveryId {
 /// Panics if the highest bit of byte at index 32 is set, as this indicates non-normalized
 /// signature. Panics if the recovery id is in reduced-x form.
 pub fn encode_signature(mut signature: [u8; 64], recovery_id: RecoveryId) -> [u8; 64] {
+    // This assertion is hit during fuzzing. Verify it is safe to disable.
+    #[cfg(not(any(fuzzing, feature = "test-helpers")))]
```
Comment (maxammann) [inline fuel-crypto/src/secp256/signature_format.rs:59]: Worth investigating. Is this assertion expected to be violated during fuzzing?

File: fuel-tx/src/transaction/validity.rs [hunk 3]
```diff
@@ -112,6 +112,8 @@ impl Input {
                     recover_address()?
                 };
 
+                // This error is reached during fuzzing often, verify it is safe to disable it
+                #[cfg(not(any(fuzzing, feature = "test-helpers")))]
                 if owner != &recovered_address {
```
Comment (maxammann) [inline fuel-tx/src/transaction/validity.rs:117]: Worth investigating. Is this assertion expected to be violated during fuzzing?

File: fuel-vm/fuzz/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+    .collect::<Vec<u8>>();
+
+    let contract_id = test_context
+        .setup_contract_bytes(subcontract.clone(), None, None)
+        .contract_id;
+
+    let max_program_length = 2usize.pow(18)
+        - test_context.get_tx_params().tx_offset()
+        - <fuel_vm::prelude::Script as Script>::script_offset_static()
+        - 256; // TODO reevalute this one
```
Comment (maxammann) [inline fuel-vm/fuzz/src/lib.rs:118]: I'm not sure where this magic value comes from, but I guess it is oke to have that.

**Comment by maxammann** [comment]
@xgreenx Let me know how you want the corpus to be delivered :) I suggest an external repository that can be public or private. Private might be better to start with.

I wanted to experiment with integrating a CI fuzzing, but that can follow up as separate PR.

File: fuel-vm/fuzz/Cargo.toml [hunk 5]
```diff
@@ -3,22 +3,50 @@ name = "fuel-vm-fuzz"
...
+name = "seed"
+path = "src/seed.rs"
+
+[[bin]]
+name = "execute"
+path = "src/execute.rs"
+
+[[bin]]
+name = "collect"
+path = "src/collect.rs"
```
Comment (xgreenx) [inline fuel-vm/fuzz/Cargo.toml:51]: It is not clear what is the purpose of each binary, could you add some documentation(maybe README file) on how to use it=) We would appreciate that

File: fuel-tx/src/transaction/validity.rs [hunk 6]
```diff
@@ -112,6 +112,8 @@ impl Input {
                     recover_address()?
                 };
 
+                // This error is reached during fuzzing often, verify it is safe to disable it
+                #[cfg(not(any(fuzzing, feature = "test-helpers")))]
```
Comment (xgreenx) [inline fuel-tx/src/transaction/validity.rs:116]: I'm curious what problem you had here and how to reproduce it? Maybe we will have some ideas on how to address that in a better way

**Review by xgreenx — COMMENTED** [review]
Just an overall question: do you have any ideas on how it can be integrated into our CI on a daily basis?

**Comment by maxammann** [comment]
> Just an overall question: do you have any ideas on how it can be integrated into our CI on a daily basis?

My idea was to experiment with https://google.github.io/clusterfuzzlite/ to add CI fuzzing in a future PR.

File: fuel-tx/src/transaction/validity.rs (hunk 6 above)
Comment (maxammann) [reply fuel-tx/src/transaction/validity.rs:116]: I don't have one at hand anymore, but removing the line and running the fuzzer should uncover it quickly. We can simply remove that for now and re-add it later on when you have CI fuzzing.

File: fuel-vm/fuzz/Cargo.toml (hunk 5 above)
Comment (maxammann) [reply fuel-vm/fuzz/Cargo.toml:51]: Added

File: fuel-tx/src/transaction/validity.rs (hunk 6 above)
Comment (maxammann) [inline fuel-tx/src/transaction/validity.rs:116]:
```suggestion
```

Comment (maxammann) [reply fuel-tx/src/transaction/validity.rs:116]: Commit suggestion if you agree with that approach.

**Comment by maxammann** [comment]
@xgreenx Here is the CFL PR: https://github.com/FuelLabs/fuel-vm/pull/727

**Comment by maxammann** [comment]
@xgreenx I think we want to add the no-changelog tag to this PR. Also the wasm checks seem to depend on the base branch being in the fuel-vm repository.

**Review by xgreenx — APPROVED** [review]
I will merge it to our local branch to fix CI and add small modifications=)

**Pull request merged**